test/e2e/create_test.go

```diff
@@ -160,30 +160,77 @@ func (tc *testContext) testMachineConfiguration(t *testing.T) {
+// configure the Machine. In order to simulate that case we need to wait for the MachineSet to be fully
+// provisioned and then change the key. The correct amount of nodes being configured is proof that the
+// mismatched Machine created with the mismatched key was deleted and replaced.
+// Depending on timing and configuration flakes this will either cause all Machines, or all Machines after
+// the first configured Machines to hit this scenario. This could be considered a platform-agnostic test, except for
+// vSphere where the private key is baked in the VM template, so we run it only on Azure.
+func (tc *testContext) testMachineConfigurationWhilePrivateKeyChange(t *testing.T) {
+	if tc.CloudProvider.GetType() != config.AzurePlatformType {
+		t.Skip("test disabled, exclusively runs on Azure")
+	}
+	machines, err := tc.waitForWindowsMachines(int(gc.numberOfMachineNodes), "Provisioned", false)
+	require.NoError(t, err, "error waiting for Windows Machines to be provisioned")
+
+	// Replace the known private key with a randomly generated one.
+	err = tc.createPrivateKeySecret(false)
+	require.NoError(t, err, "error replacing private key secret")
+
+	// Wait for machines to be reconciled by deletion
+	err = tc.waitForMachinesDeleted(machines.Items)
+	require.NoError(t, err, "error waiting for machines deletion after private key secret change")
+}
+
+// waitForMachinesDeleted waits for the given list of machines to be deleted
+func (tc *testContext) waitForMachinesDeleted(machinesToDelete []mapi.Machine) error {
+	if len(machinesToDelete) == 0 {
+		return nil
+	}
+	// deletion of machines takes time, no need to poll too frequently
+	deletionRetryInterval := time.Minute * 1
+	deletionTimeLimit := time.Duration(len(machinesToDelete)) * time.Minute * 10
+	log.Printf("waiting (timeout: %s) for %d Windows Machine(s) to be deleted",
+		deletionTimeLimit.String(), len(machinesToDelete))
+	// list only e2e labeled machines
+	listOptions := metav1.ListOptions{LabelSelector: clusterinfo.MachineE2ELabel + "=true"}
+	// wait retry interval first and then poll until all machines are deleted
+	return wait.PollUntilContextTimeout(context.TODO(), deletionRetryInterval, deletionTimeLimit, false,
+		func(ctx context.Context) (done bool, err error) {
+			currentMachines, err := tc.client.Machine.Machines(clusterinfo.MachineAPINamespace).List(context.TODO(),
+				listOptions)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return true, nil
+				}
+				log.Printf("machine object listing failed, retrying: %v", err)
+				return false, nil
+			}
+			for _, toDelete := range machinesToDelete {
+				for _, current := range currentMachines.Items {
+					if current.GetName() == toDelete.GetName() {
+						log.Printf("waiting for machine %s to be deleted", toDelete.GetName())
+						return false, nil
+					}
+				}
+			}
+			// all machines were deleted
+			return true, nil
```

Comment:
This approach is somewhat over-engineering things.
You can check if a Machine has been deleted by
```
func (tc* testContext) machineDeleted(name string) bool {
  currentMachines, err := tc.client.Machine.Machines(clusterinfo.MachineAPINamespace).Get(context.TODO(), name, getOptions)
  return apierrors.IsNotFound(err)
}
``` 
You can loop this for all the machines you are waiting for.

Your approach makes 1 api call instead of N api calls, but also complicates the code. Consider that in the context this function is used, this is:
1)  test code, so it doesn't need to be as performant as code within WMCO
2) When the tests are running, 100% of the time in CI the amount of Machines to wait for will be 1. And most of the time when ran by developers it will be 1, with the occasional 0 or 2.